File: Resources/ResourceManager.cpp
#include "ResourceManager.h"
#include "Renderer/Renderer.h"   // full type needed here (forward-declared in the header)

ResourceManager::ResourceManager(Renderer& renderer) : renderer(renderer) {}

MeshHandle ResourceManager::createMesh(const std::vector<Vertex>& vertices,
                                       const std::vector<uint32_t>& indices)
{
	uint32_t id = (uint32_t)meshes.size();
	meshes.push_back(std::make_unique<Mesh>(renderer.getDevice(), vertices, indices));
	return MeshHandle{ id };
}

TextureHandle ResourceManager::loadTexture(const std::string& path)
{
	auto tex = std::make_unique<Texture>(renderer.getDevice(), path.c_str());
	uint32_t slot = renderer.registerBindlessTexture(tex->view(), tex->sampler());
	textures.push_back(std::move(tex));
	return TextureHandle{ slot };
}

void ResourceManager::registerFlipbook(TextureHandle h, float frameCount, float flipRate)
{
	renderer.setFlipbook(h.id, frameCount, flipRate);
}

const Mesh& ResourceManager::getMesh(MeshHandle h) const
{
	return *meshes[h.id];
}

File: csharp/Resources/ResourceManager.cs
using IdleL.Rendering;

namespace IdleL.Resources;

// Owns GPU meshes + textures, handing out opaque handles. Textures register into the renderer's
// bindless array on load.
class ResourceManager : IDisposable
{
    readonly Renderer renderer;
    readonly List<Mesh> meshes = new();
    readonly List<Texture> textures = new();

    public ResourceManager(Renderer renderer) => this.renderer = renderer;

    public MeshHandle CreateMesh(ReadOnlySpan<Vertex> vertices, ReadOnlySpan<uint> indices)
    {
        uint id = (uint)meshes.Count;
        meshes.Add(new Mesh(renderer.Device, vertices, indices));
        return new MeshHandle { Id = id };
    }

    public TextureHandle LoadTexture(string path)
    {
        var tex = new Texture(renderer.Device, path);
        uint slot = renderer.RegisterBindlessTexture(tex.View, tex.Sampler);
        textures.Add(tex);
        return new TextureHandle { Id = slot };
    }

    public int TextureWidth(TextureHandle h) => textures[(int)h.Id].Width;
    public int TextureHeight(TextureHandle h) => textures[(int)h.Id].Height;
    public void RegisterFlipbook(TextureHandle h, float frameCount, float flipRate) =>
        renderer.SetFlipbook(h.Id, frameCount, flipRate);

    public Mesh GetMesh(MeshHandle h) => meshes[(int)h.Id];

    public void Dispose()
    {
        foreach (var m in meshes) m.Dispose();
        foreach (var t in textures) t.Dispose();
    }
}
